src/pyk/kcfg/store.py

```diff
@@ -0,0 +1,133 @@
+from __future__ import annotations
+
+import threading
+from abc import abstractmethod
+from collections.abc import MutableMapping
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, Generic, TypeVar, final
+
+from ..cterm import CTerm
+from ..kast.inner import KApply, KSequence, KToken, KVariable, bottom_up_with_summary
+from .kcfg import KCFG
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+    from ..kast.inner import KInner, KLabel
+
+
+A = TypeVar('A')
+
+
+@dataclass
+class CachedValues(Generic[A]):
+    value_to_id: dict[A, int] = field(default_factory=dict)
+    values: list[A] = field(default_factory=list)
```

Comment:
Consider making these private:

```suggestion
    _value_to_id: dict[A, int] = field(default_factory=dict)
    _values: list[A] = field(default_factory=list)
```

Then it can have two methods, say, `put` (currently `cache`) and `get`. The index can then be passed to `OptimizedKInner.build`.

Reply:
I added a 'get', but didn't change the `cache` name.